kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProduceAction.java

```diff
@@ -0,0 +1,363 @@
+      Collector.of(
+          ImmutableMultimap::builder,
+          (builder, header) -> builder.put(header.getName(), header.getValue()),
+          (left, right) -> left.putAll(right.build()),
+          ImmutableMultimap.Builder::build);
+
+  private final Provider<SchemaManager> schemaManager;
+  private final Provider<RecordSerializer> recordSerializer;
+  private final Provider<ProduceController> produceController;
+  private final SubjectNameStrategy defaultSubjectNameStrategy;
+
+  @Inject
+  public ProduceAction(
+      Provider<SchemaManager> schemaManager,
+      Provider<RecordSerializer> recordSerializer,
+      Provider<ProduceController> produceController,
+      SubjectNameStrategy defaultSubjectNameStrategy) {
+    this.schemaManager = requireNonNull(schemaManager);
+    this.recordSerializer = requireNonNull(recordSerializer);
+    this.produceController = requireNonNull(produceController);
+    this.defaultSubjectNameStrategy = requireNonNull(defaultSubjectNameStrategy);
+  }
+
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public void produce(
+      @Suspended AsyncResponse asyncResponse,
+      @PathParam("clusterId") String clusterId,
+      @PathParam("topicName") String topicName,
+      MappingIterator<ProduceRequest> requests) throws Exception {
+    StreamingResponse.from(requests)
+        .compose(request -> produce(clusterId, topicName, request))
+        .resume(asyncResponse);
+  }
+
+  private CompletableFuture<ProduceResponse> produce(
+      String clusterId, String topicName, ProduceRequest request) {
+    Optional<RegisteredSchema> keySchema =
+        request.getKey().flatMap(key -> getSchema(topicName, /* isKey= */ true, key));
+    Optional<EmbeddedFormat> keyFormat =
+        keySchema.map(schema -> Optional.of(schema.getFormat()))
+            .orElse(request.getKey().flatMap(ProduceRequestData::getFormat));
+    Optional<ByteString> serializedKey =
+        serialize(topicName, keyFormat, keySchema, request.getKey(), /* isKey= */ true);
+
+    Optional<RegisteredSchema> valueSchema =
+        request.getValue().flatMap(value -> getSchema(topicName, /* isKey= */ false, value));
+    Optional<EmbeddedFormat> valueFormat =
+        valueSchema.map(schema -> Optional.of(schema.getFormat()))
+            .orElse(request.getValue().flatMap(ProduceRequestData::getFormat));
+    Optional<ByteString> serializedValue =
+        serialize(topicName, valueFormat, valueSchema, request.getValue(), /* isKey= */ false);
+
+    CompletableFuture<ProduceResult> produceResult =
+        produceController.get().produce(
+            clusterId,
+            topicName,
+            request.getPartitionId(),
+            request.getHeaders().stream().collect(PRODUCE_REQUEST_HEADER_COLLECTOR),
+            serializedKey,
+            serializedValue,
+            request.getTimestamp().orElse(Instant.now()));
+
+    return produceResult.thenApply(
```

Comment:
1. The ProduceController.produce API is non-blocking.
2. Preserving the order of input/output is the whole point of `StreamingResult`. 

Initially the plan was to force idempotent producer in REST Proxy to guarantee this ordering (idempotent producer will be the default for the Java API in Kafka 3 anyways). But I didn't want to force that, since it would make Kafka REST configuration more complex that what it already is. So I implemented the ordering outside the Producer, in `StreamingResult`.